_posts/2021-05-06-broker-tuning.md

```diff
@@ -0,0 +1,443 @@
+
+The `auto.create.topics.enable` property is enabled by default to create topics automatically, but it's usually disabled as Kafka users tend to prefer applying more control over topic creation.
+If you do use automatic topic creation, set `num.partitions` to equal the number of brokers in the cluster so that writes are distributed.
+The `delete.topic.enable` property is enabled by default to allow topics to be deleted and is usually disabled too. This time you're making sure that data is not accidentally lost, although you can temporarily enable the property to delete topics if circumstances demand it.
+
+```properties
+# ...
+auto.create.topics.enable=false
+delete.topic.enable=true
+# ...
+```
+
+> You cannot delete topics with the `KafkaTopic` resource if the `delete.topic.enable` property is set to false.
+
+### Internal topic settings for transactions and commits
+
+Pay attention to the configuration requirements for internal Kafka topics.
+
+If you are [using transactions](https://strimzi.io/docs/operators/latest/using.html#reliability_guarantees) to guarantee the reliability of writes to partitions from producers, the internal `__transaction_state` topic used in the process requires a minimum of three brokers with the default settings.
+
+```properties
+# ...
+transaction.state.log.replication.factor=3
+transaction.state.log.min.isr=2
+# ...
+```
+
+The `__consumer_offsets` topic, which stores partition offset commits, has default settings for the number of partitions and replication factor.
+
+```properties
+# ...
+offsets.topic.num.partitions=50
+offsets.topic.replication.factor=3
+offsets.commit.required.acks=-1
+# ...
+```
+
+***:warning: Do not reduce these settings in production.***
+
+### Improving request handling throughput by increasing I/O threads
+
+Network threads (`num.network.threads`) handle requests to the Kafka cluster, such as produce and fetch requests from client applications.
+Adjust the number of network threads to reflect the replication factor and the levels of activity from client producers and consumers interacting with the Kafka cluster.
+
+I/O threads (`num.io.threads`) pick up requests from the request queue to process them. Adding more threads can improve throughput, but the number of CPU cores/and disk bandwidth imposes a practical upper limit.
+The minimum number of threads should equal the number of storage volumes.
+
+To reduce congestion and regulate the request traffic, you can use `queued.max.requests` to limit the number of requests allowed in the request queue before the network thread is blocked.
+Use `num.recovery.threads.per.data.dir` to specify the number of threads used for log loading at startup and flushing at shutdown.
+
+```properties
+# ...
+num.io.threads=8
+queued.max.requests=500
+num.network.threads=3
+num.recovery.threads.per.data.dir=1
+# ...
+```
+
```

Comment:
Since we've covered most of the thread pools here, it might also be worth mentioning that when reconfiguring these things """The size of various thread pools used by the broker may be updated dynamically at cluster-default level used by all brokers. Updates are restricted to the range currentSize / 2 to currentSize * 2 to ensure that config updates are handled gracefully. """

Reply:
Added:

> Configuration updates to the thread pools for all brokers might occur dynamically at the cluster level. 
> These updates are restricted to between half the current size and twice the current size.